interface/product_registration/product_registration_service.php

```diff
@@ -18,11 +18,70 @@
  * @link    http://www.open-emr.org
  */
 
+$fake_register_globals=false;
+$sanitize_all_escapes=true;
 include_once("../globals.php");
+require_once($GLOBALS['fileroot'] . "/interface/main/exceptions/invalid_email_exception.php");
+require_once($GLOBALS['fileroot'] . "/interface/main/exceptions/local_product_registration_exception.php");
+require_once($GLOBALS['fileroot'] . "/interface/main/exceptions/remote_product_registration_exception.php");
+require_once($GLOBALS['fileroot'] . "/ccr/uuid.php");
 
-// func getProductStatus() -> sqlExecute(SELECT status from product_registration LIMIT 1)
-// if opt_out is true, then return 'OPT_OUT'. If no result, return 'UNREGISTERED', if good result 'REGISTERED'
+// TODO: @Brady to help me get this in the migration DSL...
+// TODO: Figure out why foreign key returns `errno 150`... am thinking it's because user email isn't uniq.
+//CREATE TABLE product_registration (
+//    registration_id CHAR(36) NULL,
+//    email VARCHAR(255) NULL,
+//    opt_out TINYINT(1) NULL,
+//    PRIMARY KEY (registration_id)/*
+//    FOREIGN KEY (email) REFERENCES users(email)*/
+//);
 
-// func registerProduct(email) -> sqlExecute(INSERT INTO product_registration VALUES (util.uuid(), email, false))
+class ProductRegistrationService {
+    public function __construct() {}
 
-// func optOutOfProductRegistration() -> sqlExecute(INSERT INTO product_registration VALUES (null, null, false))
+    public function getProductStatus() {
+        $row = sqlQuery("SELECT * FROM product_registration LIMIT 1;");
+
+        if (empty($row)) {
+            return 'UNREGISTERED';
+        } else if (!(empty($row['registration_id']))) {
+            return 'REGISTERED';
+        } else if (!(empty($row['opt_out'])) && $row['opt_out'] == true) {
+            return 'OPT_OUT';
+        }
+    }
+
+    public function registerProduct($email) {
+        if (empty($email)) {
+            $this->optOutStrategy();
+            return null;
+        } else {
+            return $this->optInStrategy($email);
+        }
+    }
+
+    private function optInStrategy($email) {
+        if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+            throw new InvalidEmailException($email);
+        }
+
+        $uuid = getUuid();
+
+        // TODO: Is there support for parameterization for security purposes?
```

Comment:
yes, there definitely is:
``````
$result = sqlStatement("INSERT INTO product_registration VALUES (?, ?, false)", array($uuid,$email));
``````
(also since you never know where the table is going to head, should explicitly list the sql-columns that are going to be populate; INTO (...) product_registration VALUES (...) )